pkg/activator/net/throttler.go

```diff
@@ -0,0 +1,489 @@
+	// First, see if we can succeed with just an RLock. This is in the request path so optimizing
+	// for this case is important
+	t.revisionThrottlersMutex.RLock()
+	revThrottler, ok := t.revisionThrottlers[revID]
+	t.revisionThrottlersMutex.RUnlock()
+	if ok {
+		return revThrottler, nil
+	}
+
+	// Redo with a write lock since we failed the first time and may need to create
+	t.revisionThrottlersMutex.Lock()
+	defer t.revisionThrottlersMutex.Unlock()
+	revThrottler, ok = t.revisionThrottlers[revID]
+	if !ok {
+		rev, err := t.revisionLister.Revisions(revID.Namespace).Get(revID.Name)
+		if err != nil {
+			return nil, err
+		}
+		revThrottler = newRevisionThrottler(revID, rev.Spec.GetContainerConcurrency(), t.breakerParams, t.logger)
+		t.revisionThrottlers[revID] = revThrottler
+	}
+	return revThrottler, nil
+}
+
+// revisionUpdated is used to ensure we have a backlog set up for a revision as soon as it is created
+// rather than erroring with revision not found until a networking probe succeeds
+func (t *Throttler) revisionUpdated(obj interface{}) {
+	rev := obj.(*v1alpha1.Revision)
+	revID := types.NamespacedName{rev.Namespace, rev.Name}
+	t.logger.Debugf("Revision update %q", revID.String())
+
+	if _, err := t.getOrCreateRevisionThrottler(revID); err != nil {
+		t.logger.Errorw("Failed to get revision throttler for revision "+revID.String(), zap.Error(err))
+	}
+}
+
+// revisionDeleted is to clean up revision throttlers after a revision is deleted to prevent unbounded
+// memory growth
+func (t *Throttler) revisionDeleted(obj interface{}) {
+	rev := obj.(*v1alpha1.Revision)
+	revID := types.NamespacedName{rev.Namespace, rev.Name}
+	t.logger.Debugf("Revision delete %q", revID.String())
+
+	t.revisionThrottlersMutex.Lock()
+	defer t.revisionThrottlersMutex.Unlock()
+	delete(t.revisionThrottlers, revID)
+}
+
+func (t *Throttler) handleUpdate(update *RevisionDestsUpdate) {
+	if rt, err := t.getOrCreateRevisionThrottler(update.Rev); err != nil {
+		t.logger.Errorw(fmt.Sprintf("Failed to get revision throttler for revision %q", update.Rev.String()),
+			zap.Error(err))
+	} else {
+		rt.handleUpdate(t, update)
+	}
+}
+
+func (t *Throttler) updateAllThrottlerCapacity(activatorCount int) {
+	t.logger.Debugf("Updating activator count to %d.", activatorCount)
+	t.revisionThrottlersMutex.RLock()
+	defer t.revisionThrottlersMutex.RUnlock()
+
+	for _, rt := range t.revisionThrottlers {
+		t.logger.Debugf("Updating rt %v with backend count %d", rt, rt.backendCount)
+		rt.updateCapacity(t, rt.backendCount)
+	}
+}
+
+func (t *Throttler) activatorEndpointsUpdated(newObj interface{}) {
+	endpoints := newObj.(*corev1.Endpoints)
+
+	func() {
+		t.numActivatorsMux.Lock()
+		defer t.numActivatorsMux.Unlock()
+		t.numActivators = resources.ReadyAddressCount(endpoints)
+	}()
+	t.updateAllThrottlerCapacity(t.numActivators)
+}
+
+func (t *Throttler) activatorCount() int {
+	t.numActivatorsMux.Lock()
```

Comment:
RLock/RUnlock.

Also I think we can switch to `sync/atomic` for `numActivators` without loss of fidelity.